### PR TITLE
feat(route53resolver): DNS Firewall

### PR DESCRIPTION
**Pull request by jogold**
Add L2s for `FirewallDomainList`, `FirewallRuleGroup` and
`FirewallRuleGroupAssociation`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jogold** [comment]
hey @njlynch, i see you removed your assignment here, is someone picking this up? thx

**Comment by njlynch** [comment]
Hi @jogold !

>  is someone picking this up?

Not immediately. Apologies, I should have left a comment along with my unassignment to explain.

We're taking efforts to triage and prioritize issues and PRs more uniformly, and then use a combination of inputs (e.g., :+1:s, priority, comments) to decide when to work on both. In tandem, we're also trying to signal and track better what we're actively working on (thus removing my assignment).

I can't give a fixed ETA for when I'll be able to pick this up, but I'll try to carve out time in the next couple weeks. In the meantime, getting folks in the community to :+1: this will help prioritize to show there's a strong interest for these L2s.

File: packages/@aws-cdk/aws-route53resolver/lib/firewall-rule-group.ts [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+    this.rules = props.rules ?? [];
+
+    const ruleGroup = new CfnFirewallRuleGroup(this, 'Resource', {
+      name: props.name,
+      firewallRules: Lazy.any({ produce: () => this.rules.map(renderRule) }),
+    });
+
+    this.firewallRuleGroupId = ruleGroup.attrId;
+    this.firewallRuleGroupArn= ruleGroup.attrArn;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-route53resolver/lib/firewall-rule-group.ts:237]:
```suggestion
    this.firewallRuleGroupArn = ruleGroup.attrArn;
```

**Review by njlynch — CHANGES_REQUESTED** [review]
Looks great! One error message to update and two minor nitpicks. Otherwise is good to go.

Thanks!

File: packages/@aws-cdk/aws-route53resolver/lib/firewall-domain-list.ts [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+export abstract class FirewallDomains {
+  /**
+   * Firewall domains created from a list of domains
+   *
+   * @param list the list of domains
+   */
+  public static fromList(list: string[]): FirewallDomains {
+    for (const domain of list) {
+      if (!/^[\w-.]{1,128}$/.test(domain)) {
+        throw new Error(`Invalid domain: ${domain}. The name must have 1-128 characters. Valid characters: A-Z, a-z, 0-9, _, -, .`);
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-route53resolver/lib/firewall-domain-list.ts:49]: Underscores are unusual but allowed in (sub)domains names and `\w` includes underscores (`\w = [a-zA-Z0-9_]`)

https://stackoverflow.com/questions/2180465/can-domain-name-subdomains-have-an-underscore-in-it

[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-route53resolver/lib/firewall-domain-list.ts]: Removed the `{1,128}` part which was incorrect and actually applies to the domain list name.

**Comment by jogold** [comment]
@njlynch can you check this PR again? thx!

**Review by njlynch — APPROVED** [review]

**Pull request merged**